witness/internal/witness/db.go

```diff
@@ -53,8 +54,23 @@ func (d *Database) GetLatest(logPK string) (*Chkpt, error) {
 	return &maxChkpt, nil
 }
 
-// SetCheckpoint writes the checkpoint to the DB for a given logPK.
-func (d *Database) SetCheckpoint(logPK string, c Chkpt) error {
-	_, err := d.db.Exec("INSERT OR IGNORE INTO chkpts (key, size, raw) VALUES (?, ?, ?)", logPK, c.Size, c.Raw)
-	return err
+// SetCheckpoint writes the checkpoint to the DB for a given logPK, assuming
+// that the latest checkpoint is still what the caller thought it was.
+func (d *Database) SetCheckpoint(logPK string, latest, c *Chkpt) error {
+	tx, err := d.db.BeginTx(context.Background(), nil)
+	if err != nil {
+		return fmt.Errorf("BeginTx: %v", err)
+	}
+	if latest != nil {
```

Comment:
This check leaves the door open for an error in calling code to cause a witness split view. If `nil` is always passed in for `latest` then no checking will be performed. My expectation was that `latest=nil` would still cause this code to check for the latest, but enforce that it isn't present. Maybe this is a feature and not a bug, but it should definitely be documented clearly.

Reply:
Good point!  The code now checks to make sure there is nothing there if `latest=nil`, and there's a test to capture this path as well.